pkg/query/arrow_test.go

```diff
@@ -225,3 +227,233 @@ func TestArrowIteratorMultipleResources(t *testing.T) {
 	}
 	require.Equal(expected, rels)
 }
+
+func TestArrowIteratorCaveatCombination(t *testing.T) {
+	t.Parallel()
+
+	require := require.New(t)
+
+	// Create test context
+	ctx := &Context{
+		Context:  t.Context(),
+		Executor: LocalExecutor{},
+	}
+
+	t.Run("CombineTwoCaveats_AND_Logic", func(t *testing.T) {
+		t.Parallel()
+
+		// Left side path with caveat
+		leftPath := MustPathFromString("document:doc1#parent@folder:folder1")
+		leftPath.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "left_caveat",
+				},
+			},
+		}
+
+		// Right side path with different caveat (matching the left side's target)
+		rightPath := MustPathFromString("folder:folder1#viewer@user:alice")
+		rightPath.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "right_caveat",
+				},
+			},
+		}
+
+		leftIter := NewFixedIterator(leftPath)
+		rightIter := NewFixedIterator(rightPath)
+
+		arrow := NewArrow(leftIter, rightIter)
+
+		relSeq, err := ctx.Check(arrow, NewObjects("document", "doc1"), NewObject("user", "alice").WithEllipses())
+		require.NoError(err)
+
+		rels, err := CollectAll(relSeq)
+		require.NoError(err)
+
+		require.Len(rels, 1, "Arrow should return one combined relation")
+		require.NotNil(rels[0].Caveat, "Result should have combined caveat")
+		// The combination logic should combine both caveats with AND logic
+	})
+
+	t.Run("LeftCaveat_Right_NoCaveat", func(t *testing.T) {
+		t.Parallel()
+
+		// Left side path with caveat
+		leftPath := MustPathFromString("document:doc1#parent@folder:folder1")
+		leftPath.Caveat = &core.CaveatExpression{
+			OperationOrCaveat: &core.CaveatExpression_Caveat{
+				Caveat: &core.ContextualizedCaveat{
+					CaveatName: "left_caveat",
+				},
+			},
+		}
+
+		// Right side path with no caveat
+		rightPath := MustPathFromString("folder:folder1#viewer@user:alice")
+
+		leftIter := NewFixedIterator(leftPath)
+		rightIter := NewFixedIterator(rightPath)
+
+		arrow := NewArrow(leftIter, rightIter)
+
+		relSeq, err := ctx.Check(arrow, NewObjects("document", "doc1"), NewObject("user", "alice").WithEllipses())
+		require.NoError(err)
+
+		rels, err := CollectAll(relSeq)
+		require.NoError(err)
+
+		require.Len(rels, 1, "Arrow should return one relation")
+		require.NotNil(rels[0].Caveat, "Left caveat should be preserved")
+		// Note: Checking exact caveat content would require parsing the CaveatExpression
```

Comment:
Is this future work? Is it stuff that we have the tools for and it'd just be painful, or is this a project unto itself?